File: Linked_list.py
class LinkedList:

# ----------------------------------------
    class Node:
        def __init__(self, data):
            self.data = data
            self.next = None
# ----------------------------------------

    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0
# ----------------------------------------

    def addLast(self, data):
        self.node = self.Node(data)

        if self.first == None:
            self.first = self.last = self.node
        else:
            self.last.next = self.node
            self.last = self.node

        self.size = self.size + 1
# ----------------------------------------

    def addFirst(self, data):
        self.node = self.Node(data)

        if self._isEmpty():
            self.first = self.last = self.node
        else:
            self.node.next = self.first
            self.first = self.node

        self.size = self.size + 1
# ----------------------------------------

    def addMiddle(self, data, after_node):
        current = self.first
        while(current is not None):
            if (current.data == after_node):
                after_node = current
                break
            current = current.next

        Next = after_node.next

        self.node = self.Node(data)

        if self._isEmpty():
            self.first = self.last = self.node
        else:
            after_node.next = self.node
            self.node.next = Next

        self.size = self.size + 1
# ----------------------------------------

    def indexOf(self, data):
        index = 0
        current = self.first

        while(current is not None):
            if (current.data == data):
                return index
            current = current.next
            index = index + 1

        return -1
# ----------------------------------------

    def removeFirst(self):
        if self._isEmpty():
            return False

        if self.first == self.last:
            self.first = self.last = None
            return

        second = self.first.next
        self.first.next = None  # Remove the link, cleanup.
        self.first = second

        self.size = self.size - 1
# ----------------------------------------

    def removeLast(self):
        if self._isEmpty():
            return False

        if (self.first == self.last):
            self.first = self.last = None
            return

        previous = self.getPrevious(self.last)
        self.last = previous
        # now last is previous, so next line is equivalent to previous.next = None
        self.last.next = None

        self.size = self.size - 1
# ----------------------------------------

    def Size(self):
        return self.size
# ----------------------------------------

    def getPrevious(self, node):
        current = self.first

        while(current != None):
            if (current.next == node):
                return current
            current = current.next

        return None
# ----------------------------------------

    def _isEmpty(self):
        return self.first is None
# ----------------------------------------

    def toArray(self):
        array = []
        current = self.first
        while(current != None):
            array.append(current.data)
            current = current.next
        return array
# ----------------------------------------

    def _isEmpty(self):
        return self.first is None
# ----------------------------------------

    def contains(self, data):
        return self.indexOf(data) != -1
# ----------------------------------------

    def reverse(self):
        if self._isEmpty():
            return False

        current = self.first.next
        previous = self.first

        while(current != None):
            next = current.next
            current.next = previous
            previous = current
            current = next

        self.last = self.first
        self.last.next = None
        self.first = previous


mylist = LinkedList()
mylist.addFirst(10)
mylist.addLast(20)
mylist.addLast(30)
mylist.addMiddle(100, 10)
mylist.reverse()
array = mylist.toArray()

print(array)

File: Linked_list.cpp
#include <iostream>
#include <string>
#include <list>

// using namespace std;
using std::string;
using std::list;

// function signatures
string arrayToString(list<int> array);


class Linked_list {

    private:

        class Node{

            public:

                int data;
                Node *next;

                Node(int data){
                    this -> data = data;
                    this -> next = NULL;
                }
        };

        Node *first = NULL;
        Node *last = NULL;

        int size;
        
    public:

        // Constructor
        Linked_list(){
            first = NULL;
            last = NULL;
            size = 0;
        }


        void addFirst(int data){
            Node *node = new Node(data);

            if ( size == 0 )
                first = last = node;
            else
            {
                (*node).next = first;
                first = node;
            }

            size = size + 1;
            
        }

        void addLast(int data){
            Node *node = new Node(data);

            if (size == 0)
                first = last = node;
            else
            {
                (*last).next = node;
                last = node;
            }

            size = size + 1;
        }

        list<int> toArray(){
            list <int> array;

            Node *current = first;

            while ( current != NULL ){
                array.push_back(current -> data);
                current = current -> next;
            }

            return array;
        }

};

int main(int argc, char const *argv[])
{
    // linked list
    Linked_list mylist;

    mylist.addFirst(10);
    mylist.addFirst(20);
    mylist.addFirst(30);
    mylist.addFirst(40);

    mylist.addLast(10);
    mylist.addLast(20);
    mylist.addLast(30);
    mylist.addLast(40);

    string items = arrayToString(mylist.toArray());
    std::cout << items;

    return 0;
}


string arrayToString(list<int> array){
    string array_str;

    for (int i: array)
        array_str += std::to_string(i) + " ";

    return array_str;
}

